### PR TITLE
Fix ``abstract-class-instantiated`` false positive for ``__new__`` method.

### LINKED ISSUE #3060: False abstract-class-instantiated with pandas.ExcelWriter
Hello,

### Steps to reproduce
Run pylint on the following code:

```python
"""Lorem ipsum"""

import pandas as pd


def demo():
    """Demo for the false positive"""
    with pd.ExcelWriter("demo.xlsx") as writer:
        print(writer)
```

### Current behavior
Pylint gives: ``abstract-class-instantiated(E0110): test.py:8:9: demo: Abstract class 'ExcelWriter' with abstract methods instantiated``

### Expected behavior

No issue

### pylint --version output
- Astroid 2.2.5
- Pylint 2.3.1
- Python 3.7.4
- GCC 7.4.0

### Discussion

This ticket has originally been opened on pandas github: https://github.com/pandas-dev/pandas/issues/27634

According to the discussions there, it seems that pylint has issues with introspection of objects defining __new__

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #3060

### CODE REVIEW

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:100]: @Pierre-Sassoulas This is very fancy and a nice side effect of `codecov`.

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:81]: I don't think we should call this "correctly". What about: `new_instantiates_super`?

**Review by DanielNoord — CHANGES_REQUESTED** [review]

File: pylint/checkers/base/basic_error_checker.py
Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:81]: lovely, naming is hard :)

File: pylint/checkers/base/basic_error_checker.py
Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:100]: Is codecov not taking into account functional tests? Because if I add a breakpoint right before the `continue` in ln 100 my functional tests locally do stop, so I think it is covered.

Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:100]: same with ln 104, it is covered by functional tests

Comment (Pierre-Sassoulas) [reply pylint/checkers/base/basic_error_checker.py:98]: Nice tooltip ! Is this a bug and really is covered though ? Maybe it depends on the interpreter being used ? What do you have locally @clavedeluna ?

Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:98]: Ah good call, testing with python3.8 it's covered, but with 3.11 it isn't, so I assume codecov is running with the latter? Not sure how to verify that in codecov settings, it's not specified in the yaml

Comment (DanielNoord) [reply pylint/checkers/base/basic_error_checker.py:98]: It feels weird that `coverage` doesn't see this. I would need to investigate this as I have never seen such things in other projects that use `coverage`.

Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:98]: Locally running pytest with coverage I can see the following

```
pylint/checkers/base/basic_error_checker.py                    242     10    96%   284, 434-435, 489-496, 603-604
```

which tells me the lines codecov is complaining about (L98, L102) are indeed covered by tests.

Happy to continue looking into this but I would need to find this data in the action run, which I don't. So I'll block this PR for now.

**Comment by clavedeluna** [comment]
Blocked by codecov issue

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [reply pylint/checkers/base/basic_error_checker.py:98]: I never saw the edited comment. We don't install `pandas` so this indeed uncovered in the tests. `codecov` is right here.
The tests should work even without `pandas` in the environment as we don't have that as a test requirement.

Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:98]: Aha! You found the issue, that's indeed the case. The only functional test that passes thru that code is the one using pandas. I'll add one to fix this. Thanks!

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:89]: This should probably be done in a `try...except`. Or with `safe_infer`.

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:108]: This is the other way around from the docstring right? The function should probably be renamed.

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:498]: Let's make L498 a `if not` and then return. There is no need for this indentation here.

File: tests/functional/a/abstract/abstract_class_instantiated.py
Comment (DanielNoord) [inline tests/functional/a/abstract/abstract_class_instantiated.py:13]: Let's remove this test as we don't install `pandas`.

File: tests/functional/a/abstract/abstract_class_instantiated.py
Comment (DanielNoord) [inline tests/functional/a/abstract/abstract_class_instantiated.py:203]: What about:

```python
def __new__(self):
    cls = super().__new__()
    return cls

def __new__(self):
    cls = super().__new__()
    return "1"
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks for fixing the coverage and test!

File: tests/functional/a/abstract/abstract_class_instantiated.py
Comment (clavedeluna) [reply tests/functional/a/abstract/abstract_class_instantiated.py:203]: as long as the class has abstract methods, yes msg will be emitted. will add them as test cases

File: pylint/checkers/base/basic_error_checker.py
Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:89]:
```
    try:
        new = next(node.igetattr("__new__"))
    except astroid.InferenceError:
        return False
```

is the closest I see to other repo code using igetattr. However, I can't think of a single example where we'd actually raise `InferenceError` since __new__ is a dunder for pretty much any class. Do you have a suggestion? I can remove the try/except or ignore coverage for this line just for precaution.

Comment (DanielNoord) [reply pylint/checkers/base/basic_error_checker.py:89]: Hmm.. @Pierre-Sassoulas what do you want? Non-covered code? Or code that is known to raise a certain exception that we just don't know the edge case for yet?

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:501]: Can you add a test for this?

**Review by DanielNoord — CHANGES_REQUESTED** [review]

File: pylint/checkers/base/basic_error_checker.py
Comment (clavedeluna) [reply pylint/checkers/base/basic_error_checker.py:501]: Turns out I can't come up with a single class example where we would land in this code, so going to remove this check. The previous if-checks above this handle the edge cases.

**Comment by clavedeluna** [comment]
hm need to see what's up with that new primer output

**Comment by clavedeluna** [comment]
I'm still stumped by 
```
abstract-class-instantiated:
Abstract class 'DontReadFromInput' with abstract methods instantiated
https://github.com/pytest-dev/pytest/blob/ca40380e99c2cdaab1d0c041f9f28cff37ef8ff9/src/_pytest/capture.py#L344
```

I tried to reproduce it locally bit that msg doesn't get raised. Maybe a maintainer has a better idea?

**Comment by DanielNoord** [comment]
Have you checked wether `TextIO` has any "abstract" methods?

**Comment by clavedeluna** [comment]
> Have you checked wether `TextIO` has any "abstract" methods?

It [does](https://github.com/python/cpython/blob/main/Lib/typing.py#L3273) ... while I can see `DontReadFromInput` getting some `abstract-method` msg raised, I'm not sure why we'd raise `abstract-class-instantiated`? Maybe I'm missing something subtle and the primer is right.

**Comment by DanielNoord** [comment]
> > Have you checked wether `TextIO` has any "abstract" methods?
> 
> It [does](https://github.com/python/cpython/blob/main/Lib/typing.py#L3273) ... while I can see `DontReadFromInput` getting some `abstract-method` msg raised, I'm not sure why we'd raise `abstract-class-instantiated`? Maybe I'm missing something subtle and the primer is right.

I don't understand this comment. If `DontReadFromInput` has abstract methods that don't get implemented than it is correct that we raise this here as well right?

**Comment by clavedeluna** [comment]
Maybe I'm misunderstanding the purpose of the msg `abstract-class-instantiated` versus msg `abstract-method`? What I understand from the docs is that the former is telling a user "don't directly instantiate a class with `abc.ABCMeta` and with abstract-methods", but if the class instantiated is inheriting from one of these, but the instantiated class DOES implement the abstract methods, as in the case of [DontReadFromInput](https://github.com/pytest-dev/pytest/blob/ca40380e99c2cdaab1d0c041f9f28cff37ef8ff9/src/_pytest/capture.py#L196) then it's fine.

Anyway, if I'm wrong, then this PR should be able to be merged, I think everything asked is complete?

**Comment by DanielNoord** [comment]
> Maybe I'm misunderstanding the purpose of the msg `abstract-class-instantiated` versus msg `abstract-method`? What I understand from the docs is that the former is telling a user "don't directly instantiate a class with `abc.ABCMeta` and with abstract-methods", but if the class instantiated is inheriting from one of these, but the instantiated class DOES implement the abstract methods, as in the case of [DontReadFromInput](https://github.com/pytest-dev/pytest/blob/ca40380e99c2cdaab1d0c041f9f28cff37ef8ff9/src/_pytest/capture.py#L196) then it's fine.
> 
> Anyway, if I'm wrong, then this PR should be able to be merged, I think everything asked is complete?

But isn't one of the abstract methods of `DontReadFromInput` missing from the final implementation?

**Pull request closed without merging**
